### PR TITLE
Add visual mode mapping to set selection as $TM_SELECTED_TEXT

### LINKED ISSUE #49: Visual mode snippet expansion
I'm currently migrating from UltiSnips.
It is possible to use `$VISUAL` there which is the text which was selected when triggering the snippet:
You select text, type an expansion keystroke, type the snippet and then again the keystroke (in insert)
Then the snippet is expanded and every appearance of `$VISUAL` is replaced. Not literally - using `VISUAL` as a placeholder was supported as well.

Reading the docs from Microsoft `$TM_SELECTED_TEXT` should be the equivalent. But I haven't got it to work yet.
Is this supported yet?

### PR DESCRIPTION
**Pull request by jandamm**
This PR allows to add a mapping which will store the current visual selection as `$TM_SELECTED_TEXT` in the next expanded snippet.

This Closes #49 `a)`

Example:
```json
{
  "snippet": {
    "prefix": ["au"],
    "body": [ "autocmd $TM_SELECTED_TEXT" ],
    "description": "Example for selected text"
  }
}
```
```vim
xmap        <C-l>  <Plug>(vsnip-set-selected-text)
imap <expr> <C-l>  vsnip#available(1)  ? '<Plug>(vsnip-expand-or-jump)' : '<C-l>'
```

- Now you can select some text in visual mode.
- Press `<C-l>`. This will store the text in vsnip, remove the text and
  place you in insert mode.
- Now type `au` and press `<C-l>` again.

The result will be `autocmd your visual selected text here`.

Caveat: `Ctrl-v` (Visual Block) is not yet supported. I'm not sure how
the text should be concatenated.

### CODE REVIEW

**Comment by hrsh7th** [comment]
Thank you for your contribution.

I think we should be better to add `<Plug>(vsnip-select)` for TM_SELECTED_TEXT (it can be usable in visual/normal/select modes).

In normal mode, the feature will be select `<cword>` instead of visual range.

What do you think about it?

**Comment by jandamm** [comment]
`<Plug>(vsnip-select)` 👍

With normal mode I'd go the full circle and use the operator pending mode.
So `<C-l>iw` would use the current word and `<C-l>2j`would use the next two lines.

I'm busy the weekend but can implement it early next week.

**Comment by hrsh7th** [comment]
Sounds good to use operator pending mode!

(I not familiar to that 😂, thank you!)

**Comment by hrsh7th** [comment]
I'm wondering which name `<Plug>(vsnip-select-text)` or `<Plug>(vsnip-select)` is better...

**Comment by jandamm** [comment]
I think `select-text` sounds better.
It's clearer than only select. But both are fine.

**Comment by jandamm** [comment]
I've updated the name to `vsnip-select-text` and added mappings for `smap` and `nmap`.
I've also removed the `s` at the end of `xmap` so that all mappings reproduce the same state as they were started in.

**Comment by jandamm** [comment]
I've now added:
- visual block mode (`<C-v>`)
- linewise operator selections (e.g. `<C-l>j`)
- removing trailing newlines

I also fixed selection in visual line mode `V`.

Removing the trailing newline is what probably most users want.
Otherwise selecting linewise would always add an empty line in the snippet.

File: README.md [hunk 1]
```diff
@@ -52,12 +52,20 @@ NeoBundle 'hrsh7th/vim-vsnip-integ'
...
 smap <expr> <C-l>   vsnip#available(1)  ? '<Plug>(vsnip-expand-or-jump)' : '<C-l>'
 imap <expr> <Tab>   vsnip#available(1)  ? '<Plug>(vsnip-jump-next)'      : '<Tab>'
 smap <expr> <Tab>   vsnip#available(1)  ? '<Plug>(vsnip-jump-next)'      : '<Tab>'
 imap <expr> <S-Tab> vsnip#available(-1) ? '<Plug>(vsnip-jump-prev)'      : '<S-Tab>'
 smap <expr> <S-Tab> vsnip#available(-1) ? '<Plug>(vsnip-jump-prev)'      : '<S-Tab>'
+" Select text to use as $TM_SELECTED_TEXT in the next snippet.
+" See https://github.com/hrsh7th/vim-vsnip/pull/50
+nmap        <C-l>   <Plug>(vsnip-select-text)
+xmap        <C-l>   <Plug>(vsnip-select-text)
+smap        <C-l>   <Plug>(vsnip-select-text)
```
Comment (hrsh7th) [inline README.md:68]: Thank you.

Could you change a usage of mapping to `<Plug>(vsnip-select-text)c` instead?

I think `<Plug>(vsnip-select-text)c` is useful than current suggestion.

What do you think about it?

Comment (hrsh7th) [reply README.md:68]: I think we should suggest

```viml
nmap        <C-l>   <Plug>(vsnip-select-text)
xmap        <C-l>   <Plug>(vsnip-select-text)
smap        <C-l>   <Plug>(vsnip-select-text)
nmap        <C-j>   <Plug>(vsnip-select-text)c
xmap        <C-j>   <Plug>(vsnip-select-text)c
smap        <C-j>   <Plug>(vsnip-select-text)c
```

**Comment by hrsh7th** [comment]
Thank you for the great feature.
It definitely useful for me.

File: README.md (hunk 1 above)
Comment (jandamm) [reply README.md:68]: I use this mapping in `xmap`.

In `smap` you can just type and don't need to remove the text first.

In `nmap` we would have to provide a special mapping to reuse the text object.
E.g. `nmap <Plug>(vsnip-select-text-c)`

So I would suggest this mapping for `xmap`.
The `nmap` currently is like `y`, if you want I could implement a `c` version as well.
I'm not sure a `d` version would be useful though. But `c` is basically `d{*}a`. (So it wouldn't need much to support it as well)

Comment (jandamm) [reply README.md:68]: If you want to support something like this, I'd suggest either:

`<Plug>(vsnip-select-text-y)`
`<Plug>(vsnip-select-text-c)`
`<Plug>(vsnip-select-text-d)`

Here it easy to say how they behave. But in some cases two or all would be totally equal.

Alternatively:

`<Plug>(vsnip-select-text)`
`<Plug>(vsnip-cut-text)`

Where `cut` would be like the one with `c` and always ends in insert-mode with the text removed.

Comment (hrsh7th) [reply README.md:68]: Your opinion seems correct. I think we should suggest `<Plug>(vsnip-select-text)c` for only `xmap`.

For other modes/mappings seems good not to suggest anything.

Comment (jandamm) [reply README.md:74]: Added `<Plug>(vsnip-cut-text`) 👍

**Comment by hrsh7th** [comment]
I planning to improve README.md after merge.

The changes almost looks good to me.
Are you plan to any changes to this PR?

**Comment by jandamm** [comment]
Unless there is any issue I'm done.

I support every visual mode and `line`/`char` operator movements.
There is also `block` but as visual mode doesn't use `g@`, it doesn't need to be supported.

**Comment by hrsh7th** [comment]
I tested this in my environment.

I found the problem `<Plug>(vsnip-cut-text)` does not respect `virtualedit=all` in normal-mode.
Probably we should use `s:map` utility in `plugin/vsnip.vim`.

**Comment by jandamm** [comment]
Was the issue that not text was selected/removed?
In this case I would have to add `type == 'block'` as well.

**Comment by jandamm** [comment]
This is how I would expect it to work with `set virtualedit=all`

**Comment by hrsh7th** [comment]
Thank you for your fix.

Your last commit seems to solve my problem.

I've used the following snippet and its great fun!

```
  "brace": {
    "prefix": ["brace"],
    "body": ["\\{${1:$TM_SELECTED_TEXT}\\}"]
  },
  "paren": {
    "prefix": ["paren"],
    "body": ["(${1:$TM_SELECTED_TEXT})"]
  },
  "tag": {
    "prefix": ["tag"],
    "body": ["<$1>${2:$TM_SELECTED_TEXT}</$1>"]
  }
```

**Pull request merged**